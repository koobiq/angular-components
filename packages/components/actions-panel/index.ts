export * from './actions-panel';
export * from './actions-panel-config';
export * from './actions-panel-container';
export * from './actions-panel-ref';
export * from './module';
