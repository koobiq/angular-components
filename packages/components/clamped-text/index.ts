export * from './clamped-text';
export * from './module';
export * from './types';
