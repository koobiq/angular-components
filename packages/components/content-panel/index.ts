export * from './content-panel';
export * from './module';
export * from './resizable';
