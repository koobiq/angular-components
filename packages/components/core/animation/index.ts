export * from './animation';
export { fadeAnimation } from './fade-animations';
export { kbqSelectAnimations } from './select-animations';
