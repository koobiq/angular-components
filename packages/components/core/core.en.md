The `core` module is a foundational part of the **Koobiq** design system.  
It provides essential utilities, services, and components used across other modules in the system.
