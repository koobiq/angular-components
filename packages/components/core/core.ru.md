Модуль `core` является фундаментальной частью дизайн-системы **Koobiq**.  
Он предоставляет базовые утилиты, сервисы и компоненты, необходимые для построения и функционирования остальных модулей системы.
