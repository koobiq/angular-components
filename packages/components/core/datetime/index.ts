export * from './date-adapter';
export * from './date-formats';
