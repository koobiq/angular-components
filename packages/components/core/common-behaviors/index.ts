import { InjectionToken } from '@angular/core';

export { CanColor, CanColorCtor, KbqComponentColors, ThemePalette, mixinColor } from './color';
export { KBQ_SANITY_CHECKS, KbqCommonModule } from './common-module';
export { CanDisable, CanDisableCtor, mixinDisabled } from './disabled';
export { CanUpdateErrorState, CanUpdateErrorStateCtor, mixinErrorState } from './error-state';
export { HasTabIndex, HasTabIndexCtor, mixinTabIndex } from './tabindex';

export const KBQ_PARENT_ANIMATION_COMPONENT = new InjectionToken<any>('kbq-parent-animation-component');
