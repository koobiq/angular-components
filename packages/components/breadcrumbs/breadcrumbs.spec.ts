import { Component, DebugElement, QueryList, Type, ViewChild, ViewChildren } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { provideRouter } from '@angular/router';
import { KbqDefaultSizes } from '@koobiq/components/core';
import {
    KbqBreadcrumbItem,
    KbqBreadcrumbs,
    kbqBreadcrumbsConfigurationProvider,
    KbqBreadcrumbsSeparator,
    KbqBreadcrumbView
} from './breadcrumbs';

const createComponent = <T>(component: Type<T>, providers: any[] = []): ComponentFixture<T> => {
    TestBed.configureTestingModule({ imports: [component], providers }).compileComponents();
    const fixture = TestBed.createComponent<T>(component);
    fixture.autoDetectChanges();
    return fixture;
};

function getBreadcrumbsDebugElement(debugElement: DebugElement): DebugElement {
    return debugElement.query(By.directive(KbqBreadcrumbs));
}

function findAllBreadcrumbItems(debugElement: DebugElement): DebugElement[] {
    return debugElement.queryAll(By.css('.kbq-breadcrumb-item'));
}

function findAllCustomBreadcrumbItems(debugElement: DebugElement): DebugElement[] {
    return debugElement.queryAll(By.css('.custom-breadcrumb'));
}

function findAllCustomSeparators(debugElement: DebugElement): DebugElement[] {
    return debugElement.queryAll(By.css('.custom-separator'));
}

const customBreadcrumbsProvider = kbqBreadcrumbsConfigurationProvider({ firstItemNegativeMargin: true });

describe(KbqBreadcrumbs.name, () => {
    describe('core', () => {
        it('should have the correct number of breadcrumb items', () => {
            const fixture = createComponent(SimpleBreadcrumbs, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement, componentInstance } = fixture;
            const breadcrumbItems = findAllBreadcrumbItems(debugElement);
            expect(breadcrumbItems.length).toBe(componentInstance.items.length);
        });

        it('should apply the default size class to the breadcrumbs container', () => {
            const fixture = createComponent(SimpleBreadcrumbs, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement } = fixture;
            const { nativeElement } = getBreadcrumbsDebugElement(debugElement);
            expect(nativeElement.classList).toMatchSnapshot();
        });

        it('should update when items change', () => {
            const fixture = createComponent(SimpleBreadcrumbs, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement, componentInstance } = fixture;
            const updatedLength = componentInstance.items.push({ text: 'New Item', disabled: false });
            fixture.autoDetectChanges();
            const breadcrumbItems = findAllBreadcrumbItems(debugElement);
            expect(breadcrumbItems.length).toBe(updatedLength);
        });

        it('should apply the disabled state correctly to specific breadcrumb items', () => {
            const fixture = createComponent(SimpleBreadcrumbs, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement } = fixture;
            fixture.detectChanges();
            const disabledItem = findAllBreadcrumbItems(debugElement)[2];
            expect(disabledItem).toBeTruthy();
        });

        it('should enforce the max limit of breadcrumb items displayed', () => {
            const fixture = createComponent(SimpleBreadcrumbs, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement, componentInstance } = fixture;
            componentInstance.max = 4;
            componentInstance.items.push({ text: 'New Item1', disabled: false });
            componentInstance.items.push({ text: 'New Item2', disabled: false });
            fixture.detectChanges();
            const displayedBreadcrumbItems = findAllBreadcrumbItems(debugElement);
            expect(componentInstance.items.length).toBeGreaterThan(componentInstance.max);
            expect(displayedBreadcrumbItems.length).toBe(componentInstance.max);
        });
    });

    describe('customization', () => {
        it('should use the custom separator template', () => {
            const fixture = createComponent(BreadcrumbsCustomization, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement, componentInstance } = fixture;
            const customSeparators = findAllCustomSeparators(debugElement);
            expect(customSeparators.length).toBe(componentInstance.items.length - 1);
        });

        it('should use the custom view template', () => {
            const fixture = createComponent(BreadcrumbsCustomization, [
                provideRouter([]),
                customBreadcrumbsProvider
            ]);
            const { debugElement, componentInstance } = fixture;
            const customBreadcrumbs = findAllCustomBreadcrumbItems(debugElement);
            expect(customBreadcrumbs.length).toBe(componentInstance.items.length);
        });
    });
});

@Component({
    template: `
        <kbq-breadcrumbs [max]="max" [size]="size">
            @for (item of items; track item) {
                <kbq-breadcrumb-item [text]="item.text" [disabled]="item.disabled" />
            }
        </kbq-breadcrumbs>
    `,
    standalone: true,
    imports: [
        KbqBreadcrumbs,
        KbqBreadcrumbItem,
        KbqBreadcrumbsSeparator,
        KbqBreadcrumbView
    ]
})
class SimpleBreadcrumbs {
    @ViewChild(KbqBreadcrumbs) breadcrumbs: KbqBreadcrumbs;

    max: number | null = null;
    size: KbqDefaultSizes = 'normal';
    items = [
        { text: 'Home', disabled: false },
        { text: 'Library', disabled: false },
        { text: 'Data', disabled: true }
    ];
}

@Component({
    template: `
        <kbq-breadcrumbs [max]="max" [size]="size">
            <ng-template kbqBreadcrumbsSeparator>
                <div class="custom-separator">CUSTOM_SEPARATOR</div>
            </ng-template>

            @for (item of items; track item) {
                <kbq-breadcrumb-item [text]="item.text" [disabled]="item.disabled">
                    <a class="custom-breadcrumb" *kbqBreadcrumbView>CUSTOM_BREADCRUMB_TEMPLATE</a>
                </kbq-breadcrumb-item>
            }
        </kbq-breadcrumbs>
    `,
    standalone: true,
    imports: [
        KbqBreadcrumbs,
        KbqBreadcrumbItem,
        KbqBreadcrumbsSeparator,
        KbqBreadcrumbView
    ]
})
class BreadcrumbsCustomization {
    @ViewChild(KbqBreadcrumbs) breadcrumbs: KbqBreadcrumbs;
    @ViewChildren(KbqBreadcrumbItem) breadcrumbItems: QueryList<KbqBreadcrumbItem>;

    max: number | null = null;
    size: KbqDefaultSizes = 'normal';
    items = [
        { text: 'Home', disabled: false },
        { text: 'Library', disabled: false },
        { text: 'Data', disabled: false },
        { text: 'Person', disabled: false }
    ];
}
