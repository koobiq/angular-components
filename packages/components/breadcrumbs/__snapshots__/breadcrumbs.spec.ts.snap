// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqBreadcrumbs core should apply the default size class to the breadcrumbs container 1`] = `
DOMTokenList {
  "0": "kbq-breadcrumbs",
  "1": "kbq-breadcrumbs_normal",
  "2": "kbq-breadcrumbs_first-item-negative-margin",
}
`;
